### ISSUE TITLE
Various portability issues

### ISSUE THREAD

**Issue by Rhialto** [SDL-Hercules-390/hyperion#279]
Hi! I am trying to package this version of hercules (in particular, release 4.1.2) for pkgsrc (the cross-platform packaging system, https://www.pkgsrc.org ), but I'm running into a couple of issues.

There is a packaged version of hercules 3.13 (see http://pkgsrc.se/emulators/hercules and http://cvsweb.netbsd.org/bsdweb.cgi/pkgsrc/emulators/hercules)

The issues I have identified so far include:

- The configure script uses `test xxx == yyy` where it should use `=` instead (several times)
- Some conditional compilation should check for `__NetBSD__` and usually it should go the same way as `__FreeBSD__`
- `pthread_setname_np()` is called with too few arguments (needs an extra argument)
- `#define int32_t                 int32_t` is a re-definition of `int32_t`, and I don't think a program is to use this reserved name in this way at all
- `tuntap.c` is totally unsuitable for NetBSD, I had to patch it out completely (like SOLARIS)
- and then the big whammo (I can try to patch around all the rest, but not this one): libtool complains about linking shared objects with static libraries isn't a proper thing to do.

```
*** Warning: Trying to link with static lib archive ./decNumber/lib/libdecNumber64.a.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because the file extensions .a of this argument makes me believe
*** that it is just a static archive that I should not used here.
````
and that 4 times for 4 different `.a` files.
Interestingly, when re-trying to link I didn't get those messages again but linking still failed:
```
...
./.libs/libherc.so: undefined reference to `i64_to_f128'
./.libs/libherc.so: undefined reference to `decNumberToString'
./.libs/libherc.sogmake[2]: *** [Makefile:2059: cckddiag] Error 1
: undefined reference to `decPackedToNumber'
./.libs/libherc.so: undefined reference to `f64_roundToInt'
./.libs/libherc.so: undefined reference to `f128_add'
./.libs/libherc.so: undefined reference to `i32_to_f32'
...
```
And when I looked a bit better, said static libraries turned out to be directly from the distfile, with no source included. I don't know for which CPU and operating system combination those are built, but they are not useful for anyone else. Can you please include the source for all these libraries?

You can see how far I got at https://wip.pkgsrc.org/cgi-bin/gitweb.cgi?p=pkgsrc-wip.git;a=commit;h=8950d35f793e874def78405b1945a8f7bc821438  (and probably soon at mirror https://github.com/NetBSD/pkgsrc-wip/tree/master/sdl-hercules-hyperion ).

Can you start with incorporating the portability fixes?

**Comment by Fish-Git** [comment]
> I am trying to package this version of hercules (in particular, release 4.1.2)

May I presume "4.1.2" is a simple typo? There is no version 4.1.2. There's a version "4.2.1", but version "4.1.2" does not exist. Did you mean "4.2.1"?

**Comment by Rhialto** [comment]
Eh yes, that is what I meant.

**Comment by Fish-Git** [comment]
> * The configure script uses `test xxx == yyy` where it should use `=` instead (several times)

I could find only _**one**_ instance of this erroneous syntax in our configure.ac script on line 1209:

https://github.com/SDL-Hercules-390/hyperion/blob/c67ac2ee50d899ea4002a16bdbfa35b6cc14f8b6/configure.ac#L1209-L1211

> * Some conditional compilation should check for `__NetBSD__` and usually it should go the same way as `__FreeBSD__`

_Technically_, we should not be performing _any_ platform tests at all. Testing for specific platforms is the opposite of portability. Instead, we should be testing for _features_. That way, if a given platform does not support a given feature today, but begins supporting it tomorrow, there is nothing we need to change. By introducing platform tests instead of _feature_ tests, you make your product _less_ portable and more fragile.

At one time we did not have any platform tests at all (except for Windows which is a special case), but over the years they have unfortunately snuck in. Now it's a mess, and is going to be very difficult to straighten out.

Adding tests for __NetBSD__ as you suggested would only compound the problem we already have.

> * `pthread_setname_np()` is called with too few arguments (needs an extra argument)

Please show me your configure output and your `config.h` file, because this should not be!

Our `configure.ac` script is checking for both the availability of the `pthread_setname_np` function itself as well as whether or not it needs one or two arguments, causing Hercules to build itself accordingly:

https://github.com/SDL-Hercules-390/hyperion/blob/c67ac2ee50d899ea4002a16bdbfa35b6cc14f8b6/configure.ac#L546-L547

https://github.com/SDL-Hercules-390/hyperion/blob/c67ac2ee50d899ea4002a16bdbfa35b6cc14f8b6/configure.ac#L1015-L1017

https://github.com/SDL-Hercules-390/hyperion/blob/c67ac2ee50d899ea4002a16bdbfa35b6cc14f8b6/configure.ac#L1847-L1870

https://github.com/SDL-Hercules-390/hyperion/blob/c67ac2ee50d899ea4002a16bdbfa35b6cc14f8b6/configure.ac#L2869-L2870

https://github.com/SDL-Hercules-390/hyperion/blob/c67ac2ee50d899ea4002a16bdbfa35b6cc14f8b6/hmacros.h#L486-L499

https://github.com/SDL-Hercules-390/hyperion/blob/c67ac2ee50d899ea4002a16bdbfa35b6cc14f8b6/hscutl.c#L1643-L1670

> * `#define int32_t                 int32_t` is a re-definition of `int32_t`, and I don't think a program is to use this reserved name in this way at all

One of our external packages needs `int32_t` (probably [SoftFloat](https://github.com/SDL-Hercules-390/SoftFloat), but I'm too lazy to check right now to be sure), and since it can't know whether there's already a `typedef` for that type or not (and AFAIK you can't issue a `typedef` twice), it instead simply `#define`s `int32_t` to itself.

This may or may not be "poor form", I don't know, but it is [supposedly quite common](https://gcc.gnu.org/onlinedocs/cpp/Self-Referential-Macros.html) and should not, AFAIK, cause any problems.

Why does NetBSD care anyway? What error is it throwing?

> * `tuntap.c` is totally unsuitable for NetBSD, I had to patch it out completely (like SOLARIS)

Don't _even_ get me started on `tuntap`!!!! &nbsp; `>8-<`

You Linux geeks are on your own with that!!

> * and then the big whammo (I can try to patch around all the rest, but not this one): libtool complains about linking shared objects with static libraries isn't a proper thing to do.

This is a mostly completely bogus warning. The libtool you are using is presuming it shouldn't link with a `.a` static library because most static libraries use position _dependent_ code and not position _independent_ code like shared libraries are required to do.

Since all of our [external packages](https://github.com/SDL-Hercules-390) ([crypto](https://github.com/SDL-Hercules-390/crypto), [decNumber](https://github.com/SDL-Hercules-390/decNumber), [SoftFloat](https://github.com/SDL-Hercules-390/SoftFloat) and [telnet](https://github.com/SDL-Hercules-390/telnet)) are all always built with the `-fPIC` option however, the waning can be safely ignored.

As far as your undefined reference link errors are concerned, did you build all of the external packages before attempting to build Hercules (and specify the `--enable-extpkgs=DIR` configure option when you did), as explained in our documented procedure for building SDL Hyperion Hercules? (i.e. step 5 of the "[Building from source - Linux and Mac OS X](https://sdl-hercules-390.github.io/html/hercinst.html#instsource)" procedure). You might need to do that. The default static libs that come delivered with SDL Hyperion Hercules are generic in nature, created on a CentOS system. Since you're trying to build on NetBSD, you probably need to build all of the external packages first, and then specify the `--enable-extpkgs=DIR` configure option when you build Hercules as previously mentioned.

**Comment by Rhialto** [comment]
On Sat 01 Feb 2020 at 09:08:23 -0800, Fish-Git wrote:
> > * The configure script uses `test xxx == yyy` where it should use `=` instead (several times)
> 
> I could find only _**one**_ instance of this erroneous syntax in our configure.ac script on line 1209:
> 
> https://github.com/SDL-Hercules-390/hyperion/blob/c67ac2ee50d899ea4002a16bdbfa35b6cc14f8b6/configure.ac#L1209-L1211

I had to search a bit for the other too:

https://github.com/SDL-Hercules-390/hyperion/blob/master/autoconf/hercules.m4#L190

> > * Some conditional compilation should check for `__NetBSD__` and usually it should go the same way as `__FreeBSD__`
> 
> _Technically_, we should not be performing _any_ platform tests at
> all. Testing for specific platforms is the opposite of portability.

Yes, I totally agree. I was basically just describing what I was doing.
Patching to make something build for a packaging system generally does
not do the restructuring that the upstream should be doing, but usually
tries to gets away with the smallest patch possible.

> > * `pthread_setname_np()` is called with too few arguments (needs an extra argument)
> 
> Please show me your configure output and your `config.h` file, because this should not be!

Interestingly, "our" declaration of `pthread_setname_np()` even required
*three* arguments, although from the description it seems like it is
really supposed to be some varargs printf-like call. So it is rather
weird, and now that you mention it, I vaguely recall reading some
discussion about this function (or one like it) where this weirdness was
called out and some improvement was proposed.

https://netbsd.gw.com/cgi-bin/man-cgi?pthread_setname_np++NetBSD-8.1
```
THREAD_GETNAME_NP(3)      Library Functions Manual      PTHREAD_GETNAME_NP(3)

SYNOPSIS
     int
     pthread_setname_np(pthread_t thread, const char *name, void *arg);

DESCRIPTION
     The pthread_setname_np() function sets the descriptive name of a thread.
     It takes the following arguments:

           thread  The thread whose descriptive name will be set.

           name    The printf(3) format string to be used to construct the
                   descriptive name of the thread.  The resulted descriptive
                   name should be shorter than PTHREAD_MAX_NAMELEN_NP.

           arg     The printf(3) argument used with name.
```
This looks like a mess..

> > * `#define int32_t                 int32_t` is a re-definition of `int32_t`, and I don't think a program is to use this reserved name in this way at all
> 
> Why does NetBSD care anyway? What error is it throwing?

I got warnings that int32_t was being re-#defined. The old value was, I
suspect, from <stdint.h>:

```
#ifndef int32_t
typedef __int32_t       int32_t;
#define int32_t         __int32_t
#endif
```

I worked around this by putting an `#ifndef int32_t` around it:
https://github.com/NetBSD/pkgsrc-wip/blob/master/sdl-hercules-hyperion/patches/patch-htypes.h

> One of our external packages needs `int32_t` (probably [SoftFloat](https://github.com/SDL-Hercules-390/SoftFloat), but I'm too lazy to check right now to be sure), and since it can't know whether there's already a `typedef` for that type or not (and AFAIK you can't issue a `typedef` twice), it instead simply `#define`s `int32_t` to itself.
> 
> This may or may not be "poor form", I don't know, but it is [supposedly quite common](https://gcc.gnu.org/onlinedocs/cpp/Self-Referential-Macros.html) and should not, AFAIK, cause any problems.

Yes, the typedef thing is a pain. Also, various C header files are
allowed to expose only *some* standard-specified names, even if in its
implementation it makes sense to use some extra ones. It seems that the
following pattern has been developed to handle this. All these
problematic names are given internal versions (like `__int32_t` above)
which are defined in some private header, taking care to prevent double
definitions. Every other header that wants to use one of them, includes
the private header. Only the public header that is supposed to defined
them, includes the private one and creates a public definition with
help of the private one. This is a bit of a pain to use systematically
but in the end it works well.

> > * `tuntap.c` is totally unsuitable for NetBSD, I had to patch it out completely (like SOLARIS)
> 
> Don't _even_ get me started on `tuntap`!!!! &nbsp; `>8-<`

Hah yes. This stuff is painfully different in all Unix variants. I
don't even want to try to look at it in the first pass. But I have some
experience with it; I added/improved tun/tap code in klh-10, a PDP-10
emulator. https://github.com/PDP-10/klh10/blob/master/src/osdnet.c
So I probably come back at this later. Is there an easy option to omit
the tun/tap support completely for the moment? I fear undefined symbols
are in my future if I just try to disable it crudely.

> Since all of our [external packages](https://github.com/SDL-Hercules-390) ([crypto](https://github.com/SDL-Hercules-390/crypto), [decNumber](https://github.com/SDL-Hercules-390/decNumber), [SoftFloat](https://github.com/SDL-Hercules-390/SoftFloat) and [telnet](https://github.com/SDL-Hercules-390/telnet)) are all always built with the `-fPIC` option however, the waning can be safely ignored.
> 
> As far as your undefined reference link errors are concerned, did you
> build all of the external packages before attempting to build Hercules
> (and specify the `--enable-extpkgs=DIR` configure option when you
> did), as explained in our documented procedure for building SDL
> Hyperion Hercules? (i.e. step 5 of the "[Building from source - Linux
> and Mac OS
> X](https://sdl-hercules-390.github.io/html/hercinst.html#instsource)"

Ahhh! I think that this explains something. I was following the
instructions from the INSTALL document at
https://github.com/SDL-Hercules-390/hyperion/blob/master/INSTALL which
are very similar, but leave out any mention of these external packages!
The BUILDING document suffers from the same omission, it seems. With so
many documents to choose from, and the Hercules 3.13 distfile containing
at least some of these sources, I can understand that I picked the wrong
one and didn't realise :)

-Olaf.
-- 
Olaf 'Rhialto' Seibert -- rhialto at falu dot nl
___  Anyone who is capable of getting themselves made President should on
\X/  no account be allowed to do the job.       --Douglas Adams, "THGTTG"

**Comment by Fish-Git** [comment]
> > * The configure script uses `test xxx == yyy` where it should use `=` instead (several times) I could find only _**one**_ instance of this erroneous syntax in our configure.ac script on line 1209: https://github.com/SDL-Hercules-390/hyperion/blob/c67ac2ee50d899ea4002a16bdbfa35b6cc14f8b6/configure.ac#L1209-L1211
> I had to search a bit for the other too: https://github.com/SDL-Hercules-390/hyperion/blob/master/autoconf/hercules.m4#L190

Thank you! I will get that fixed for you.

> > * Some conditional compilation should check for `__NetBSD__` and usually it should go the same way as `__FreeBSD__` _Technically_, we should not be performing _any_ platform tests at all. Testing for specific platforms is the opposite of portability.
> Yes, I totally agree. I was basically just describing what I was doing. Patching to make something build for a packaging system generally does not do the restructuring that the upstream should be doing, but usually tries to gets away with the smallest patch possible.

Understood. As much as I hate doing it, I will try to do what you suggest: add checks for `__NetBSD__`  everywhere `__FreeBSD__` is being checked (i.e. "if freebsd OR netbsd then...").



> Interestingly, "our" declaration of `pthread_setname_np()` even required *three* arguments, although from the description it seems like it is really supposed to be some varargs printf-like call.
> [...]
> This looks like a mess..

Yeah. &nbsp; `:(`

That's what happens when there are a zillion different distros like there are with Linux etc. With Windows it's much easier as there's one and _only_ one "Windows" from a programming API point of view. With Linux etc. however, it seems each distro has its own "better way" of doing certain things, which make supporting Linux a royal PITA! &nbsp; `:(`

(sigh..) I'll try to update `configure.ac` to detect/support the netbsd "3 arguments" case for you.

> I worked around this by putting an `#ifndef int32_t` around it: https://github.com/NetBSD/pkgsrc-wip/blob/master/sdl-hercules-hyperion/patches/patch-htypes.h

Thanks. That's a simple enough change. I'll try to get that fixed for you right away too.

And by "right away" I mean within the next few days. I'm dog tired right now and falling asleep even as I type this, so all of these changes aren't going to happen right way. Give me a few days. 'K?

> > Don't _even_ get me started on `tuntap`!!!! &nbsp; `>8-<`
> 
> Hah yes. This stuff is painfully different in all Unix variants. I don't even want to try to look at it in the first pass.

> So I probably come back at this later.

> Is there an easy option to omit the tun/tap support completely for the moment?

I don't think so, no. But let me get some much needed sleep first and then I'll poke around some and get back to you with a more definitive answer.

> Ahhh! I think that this explains something. I was following the instructions from the INSTALL document at https://github.com/SDL-Hercules-390/hyperion/blob/master/INSTALL which are very similar, but leave out any mention of these external packages!

Well it DOES have a big hard-to-miss "NOTE!" section at the very top (at the very beginning of the document) that states quite clearly that you should use both

* https://sdl-hercules-390.github.io/html/hercinst.html, and
* https://github.com/SDL-Hercules-390/hyperion/blob/master/README.md

for the authoritative instructions regarding build SDL Hyperion Hercules, and that the information in the INSTALL document only provides very generalized/generic instructions. For the specific _detailed_ build instructions you need to follow the hercinst.html web page. (The README web page provides links to all of our other README documents too in addition to one for building and installing).

So _this_ particular complaint is your own fault! &nbsp; `;-)`

Thank you for reporting all of these things, Rhialto. As I said I'll try get them all fixed for you as soon as I can.

But only after I get some much needed sleep!!!

**Comment by Rhialto** [comment]
On Sat 01 Feb 2020 at 12:59:36 -0800, Fish-Git wrote:
> But only after I get some much needed sleep!!!

Thanks, and no hurry! Sleep is very important, as I'm learning over time :)

**Comment by Fish-Git** [comment]
@Rhialto

Done! &nbsp;(I think)

Please see commits c740f710be362b5bf9e857e7423b1f21e59c344a, edf7171bfb61bb8a80860d6b8cb021f871a20fd1, 51a24502d27d76751b3e3b29d4e4b834ac5f3699 and 2ff0c0d97ca6f42d79295b865ceed5d5ca8aa4f7.

I believe that's everything.

**NOTE:** &nbsp;I did _not_ test any of this! &nbsp;Please confirm that Herc now builds okay for you.

I will wait for your confirmation first before closing this issue.

Thanks.

**Comment by wrljet** [comment]
(Not sure if this is the right place to post this, so go easy on me.)

For what it's worth, I tried building this on NetBSD 9.0_RC2 and didn't get very far.

util/bldlvlck doesn't work.
On NetBSD, Perl is installed by default in a different place than the script is expecting.  Adjusting for that, then I encountered a few of the required pgms don't use the same command line option as expected to return their version info.

Skipping ahead to ./configure.  It uses  'read -d' which isn't supported on NetBSD.

Bill

**Comment by Fish-Git** [comment]
Bill Lewis wrote:

> util/bldlvlck doesn't work. On NetBSD, Perl is installed by default in a different place than the script is expecting.

I don't understand.

I don't know perl, but looking at the `bldlvlck` script it looks to me like all it's doing is a simple `which`  command to determine whether the given facility (perl in this particular case) is installed or not:

https://github.com/SDL-Hercules-390/hyperion/blob/2ff0c0d97ca6f42d79295b865ceed5d5ca8aa4f7/util/bldlvlck#L156-L164

Are you saying that `which perl` is _failing_ on your system?  Why would that be?  Why can't `which` find perl on your system?

> Adjusting for that, I encountered a few of the required tools don't use the same command line option as expected to return their version info.

Which ones? And what is the command line option for each one to report the version?

Someone (not me! I don't know perl!) should probably rewrite this script to:

1. add a new "version opt"(column?) to the table
2. check up front (at the very beginning) what system it's running on (Apple, NetBSD, etc) on and then adjust its table entries as needed based on that.

Then the rest of the script would probably be much easier since there wouldn't be so many special case exceptions like I'm currently seeing (e.g. libtoolize, m4, perl, sed, etc).

Does anyone out there know perl?  Can you help us out please?

> Skipping ahead to `./configure`, it's using `read -d` which isn't supported on NetBSD.

_(frick!)_ I don't know &ast;nix very well at all, including bash, and received a lot of help from others when I wrote that part of `configure.ac`. I was trying so hard to be as portable as possible too. (sigh!)

So tell me, Bill: what fricking syntax does NetBSD support?! &nbsp; <code>%-P\`</code>

In any case, I _do_ want to thank you, Bill, for taking the time to report your findings. We appreciate it, and we'll try to get these issues fixed for you and Rhialto just as soon as we can.

And by "as soon as we can" I mean _eventually._

_(just please don't hold your breath!)_

&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; `;-)`

**Comment by Rhialto** [comment]
Fortunately it isn't as horrible as it may seem at first. Here is the output of the script for me (I used the release version but it is likely the same as the master version):
```
murthe.7:.../hercules/hyperion/util$ perl ./bldlvlck 
This utility will check the level of various utilities needed to build
hercules. Checking is done against versions that are KNOWN to work.
This doesn't mean a build will NOT succeed with older versions
of the utilities, but will give a hint as to what package may need
an upgrade if the build ever fails with some odd reason.


OK      autoconf requires 2.64, found 2.69

OK      automake requires 1.9, found 1.16.1

OK      flex requires 2.5, found 2.6.3

OK      gawk requires 3.0, found 5.0.1

OK      gcc requires 3, found 3.201809.5

OK      grep requires 1, found 2.5.1

m4 version 20150116
Use of uninitialized value $resp[0] in scalar chomp at ./bldlvlck line 58.
Use of uninitialized value $resp in pattern match (m//) at ./bldlvlck line 118.
Use of uninitialized value $resp in pattern match (m//) at ./bldlvlck line 122.
HUH?
OK      m4 requires 1.4.6, found 1

HUH?
UPGRADE make requires 3.79, found 1
        URL: http://www.gnu.org/directory/make.html

OK      perl requires 5.6, found 5.30.1

sed: unknown option -- -
Usage:  sed [-aElnru] command [file ...]
        sed [-aElnru] [-e command] [-f command_file] [-I[extension]]
            [-i[extension]] [file ...]
Use of uninitialized value $resp[0] in scalar chomp at ./bldlvlck line 58.
Use of uninitialized value $resp in pattern match (m//) at ./bldlvlck line 118.
Use of uninitialized value $resp in pattern match (m//) at ./bldlvlck line 122.
HUH?
UPGRADE sed requires 3.02, found 1
        URL: http://www.gnu.org/directory/sed.html
```
The tools it complains about are `m4`, `make` and `sed`. Fortunately it is documented that the GNU versions of these are to be used (`README.BSD`), and this is indeed very common. (It used to be that there was this "All the world is a VAX" laziness that non-VAX users suffered from; these days it has been replaced with "All the Unix world is Linux").

For this reason, `pkgsrc` has an easy way to supply GNU make (etc) to the build. See my work-in-progress pkgsrc entry at https://github.com/NetBSD/pkgsrc-wip/blob/4610f317fdf1d7cd7f45332a60fcb070464a6716/sdl-hercules-hyperion/Makefile#L18:
```
USE_TOOLS+=	autoconf automake flex gawk grep gm4 gmake perl gsed
```
Pkgsrc will automatically ensure that these tools are installed before starting to build this package, and all will be well.

I hope to get around this weekend to creating packages for `decNumber` etc, so that I can use them with hercules. Then I'll have to decide if I do that based on the current master, or the release plus some patches, or what, and how to avoid the tun/tap swamp for now :)

**Comment by wrljet** [comment]
> 
> Bill Lewis wrote:
> 
> > util/bldlvlck doesn't work. On NetBSD, Perl is installed by default in a different place than the script is expecting.
> 
> I don't understand.

I meant the Perl itself that runs the bldlvlck script.  It's pointed to by the shebang line:

`#!/usr/bin/perl -w`

On this NetBSD, Perl is at `/usr/local/bin/perl`

I believe one fix for that is to use:
`#! /bin/sh`
`PERLPROG=perl`
`which ${PERLPROG} > /dev/null`
`if [ "$?" -ne 0 -a -x /usr/local/bin/perl ] ; then`
`    PERLPROG=/usr/local/bin/perl`
`fi`
`exec ${PERLPROG} -w -x $0 "$@"`
`#! perl -w`

That works on NetBSD 9.0 and Debian 10 (all I've tested so far).

> 
> I don't know perl, but looking at the `bldlvlck` script it looks to me like all it's doing is a simple `which` command to determine whether the given facility (perl in this particular case) is installed or not:
> 
> https://github.com/SDL-Hercules-390/hyperion/blob/2ff0c0d97ca6f42d79295b865ceed5d5ca8aa4f7/util/bldlvlck#L156-L164
> 
> Are you saying that `which perl` is _failing_ on your system? Why would that be? Why can't `which` find perl on your system?

Nope, not -that- detection.  See above.

> 
> > Adjusting for that, I encountered a few of the required tools don't use the same command line option as expected to return their version info.
> 
> Which ones? And what is the command line option for each one to report the version?

The ones which do not work are:
- **gawk**  (just plain missing from the system, but it has `awk` which has this style of version info `"awk version 20121220"`)
- **gcc** (`gcc --version` returns `"gcc (nb3 20190319) 7.4.0"`, which the script cannot parse)
- **m4** (`m4 --version` returns `"m4 version 20150116"`, a format the script isn't expecting)
- **make** (this `make` doesn't appear to have any way to report its version info)
- **sed** (this `sed` doesn't appear to have any way to report its version info)

> 
> Someone (not me! I don't know perl!) should probably rewrite this script to:
> 
>     1. add a new "version opt"(column?) to the table
> 
>     2. check up front (at the very beginning) what system it's running on (Apple, NetBSD, etc) and then adjust its table entries accordingly based on that.

Good ideas.

> 
> Then the rest of the script would probably be much easier since there wouldn't be so many special case exceptions like I'm currently seeing (e.g. libtoolize, m4, perl, sed, etc).

The `weird()` function includes code for `libtool`, which is not in the version data table at the top.


> Does anyone out there know perl? Can you help us out please?

I could help with that.  Slowly

 
> > Skipping ahead to `./configure`, it's using `read -d` which isn't supported on NetBSD.
> 
> _(frick!)_ I don't know *nix very well at all, including bash, and received a lot of help from others when I wrote that part of `configure.ac`. I was trying so hard to be as portable as possible too. (sigh!)
> 
> So tell me, Bill: what fricking syntax does NetBSD support?!   `` %-P` ``

That I don't know yet.  I haven't looked at it.

 
> In any case, I _do_ want to thank you, Bill, for taking the time to report your findings. We appreciate it, and we'll try to get these issues fixed for you and Rhialto just as soon as we can.

This NetBSD belongs to a friend and he lets me use it to host some web pages.
I only tried this out on there to be helpful to the group.  Otherwise I know next to nothing about NetBSD, other than a lot of things are different.  Why?  Because of course they are.  
 
> 
> And by "as soon as we can" I mean _eventually._
> 
> _(just please don't hold your breath!)_
> 
>           `;-)`

Understood.  I will try to help.

Bill

**Comment by Rhialto** [comment]
Ok, some progress. Let's see if I can recall all the relevant details.

I created 4 packages for crypto, decNumber, SoftFloat and telnet. Fortunately they all build and install in the same way so I could just almost copy/paste the first one to create the rest.
What's a bit weird is how the name of the build directory is relevant. I can see how it fits with the Visual Studio build but it is of course un-Unixy, and it causes some trouble later on.
As far as I could tell, the "bitness" (32 or 64) of the build makes no difference, at least on Unix. If it were important, it is something that should be detected by cmake, not dictated to it.

I kept having trouble with libtool, which didn't want to link with the libtelnet.a (etc) files. Diving into libtool docs, it seems that this sort of thing is rather discouraged. See https://www.gnu.org/software/libtool/manual/html_node/Static-libraries.html . There is the case of "convenience libraries" built from PIC objects, but it seems that the libtool I have has rather strict ideas about when it accepts a library as such:
```
# Method to check whether dependent libraries are shared objects.
deplibs_check_method="match_pattern /lib[^/]+(\\.so|_pic\\.a)\$"
```
(and I can't even blame it very much; libfoo_pic.a is the standard name pattern in NetBSD's build system)
So I had to make sure the library wasn't named `libtelnet64.a` but `libtelnet_pic.a`. I needed to apply some blunt force to the packages to make it so, and to hercules to look for those names.
Libtool still didn't like it very much but at least did what it was told:
```
*** Warning: Linking the shared library hdt3525.la against the
*** static library /usr/pkg/lib/sdlhercules/lib/libcrypto_pic.a is not portable!
```
Then everything linked, except for an undefined reference in various places:
```
/home/tmp/wip/sdl-hercules-hyperion/work.x86_64/hyperion-Release_4.2.1/.libs/libhercu.so: undefined reference to `pthread_rwlockattr_setpshared'
```
When I looked a bit better at the manpage for that function I found
```
BUGS
     The pthread_rwlockattr_getpshared() and pthread_rwlockattr_setpshared()
     functions are hidden by default since only thread shared attributes are
     supported.
```
so I just commented the only call out. I hope it isn't a disaster.

Finally, when in the packaging phase, pkgsrc warns several times like so:
```
=> Checking file-check results for sdl-hercules-hyperion-4.2.1
ERROR: ************************************************************
ERROR: The following files are in the PLIST but not in /home/tmp/wip/sdl-hercules-hyperion/work.x86_64/.destdir/usr/pkg:
ERROR:         /home/tmp/wip/sdl-hercules-hyperion/work.x86_64/.destdir/usr/pkg/lib/hercules/dyncrypt.a
ERROR:         /home/tmp/wip/sdl-hercules-hyperion/work.x86_64/.destdir/usr/pkg/lib/hercules/dyngui.a
ERROR:         /home/tmp/wip/sdl-hercules-hyperion/work.x86_64/.destdir/usr/pkg/lib/hercules/hdt1052c.a
ERROR:         /home/tmp/wip/sdl-hercules-hyperion/work.x86_64/.destdir/usr/pkg/lib/hercules/hdt1403.a
ERROR:         /home/tmp/wip/sdl-hercules-hyperion/work.x86_64/.destdir/usr/pkg/lib/hercules/hdt2703.a
...etc...
```
(this directory is the staging install directory, where "make install" puts its stuff, so that pkgsrc can cleanly generate a nice tar file from it).

It turns out that there are a bunch of `.la` files installed that refer to corresponding `.so` files (which are installed) but also to corresponding `.a` files, which are not. In the build directory they are there, so they *are* built.

For instance `libhercu.la` contains
```
# The name of the static archive.
old_library='libhercu.a'
```

I need to find out how to clean that up; hopefully just an appropriate option such as `--no-static` or somesuch when creating these libraries.

**Comment by Rhialto** [comment]
Oh, here are the packages:
https://github.com/NetBSD/pkgsrc-wip/tree/master/sdl-hercules-crypto
https://github.com/NetBSD/pkgsrc-wip/tree/master/sdl-hercules-decnumber
https://github.com/NetBSD/pkgsrc-wip/tree/master/sdl-hercules-hyperion
https://github.com/NetBSD/pkgsrc-wip/tree/master/sdl-hercules-softfloat
https://github.com/NetBSD/pkgsrc-wip/tree/master/sdl-hercules-telnet

**Comment by Rhialto** [comment]
Gee, libtool is way too smart for its own good:
```
murthe:.../sdl-hercules-hyperion/work.x86_64/hyperion-Release_4.2.1$ ldd /home/tmp/wip/sdl-hercules-hyperion/work.x86_64/.destdir/usr/pkg/lib/hercules/hdt1052c.so
/home/tmp/wip/sdl-hercules-hyperion/work.x86_64/.destdir/usr/pkg/lib/hercules/hdt1052c.so:
        -lherc => /usr/pkg/lib/libherc.so
        -lherct => /usr/pkg/lib/libherct.so
        -lhercu => /usr/pkg/lib/libhercu.so
        -lhercs => /usr/pkg/lib/libhercs.so
        -lltdl.7 => /usr/pkg/lib/libltdl.so.7
        -lc.12 => /usr/lib/libc.so.12
        -lpthread.1 => /usr/lib/libpthread.so.1
        -lrt.1 => /usr/lib/librt.so.1
        -lz.1 => /usr/lib/libz.so.1
        -lresolv.3 => /usr/lib/libresolv.so.3
        -lm.0 => /usr/lib/libm.so.0
        -lbz2.1 => /usr/lib/libbz2.so.1
        -lhercd => /usr/pkg/lib/libhercd.so
        -ldecNumber => /usr/pkg/lib/libdecNumber.so
        -lsoftfloat => /usr/pkg/lib/libsoftfloat.so
```
so even though linking is supposed to happen against `/some/path/libdecNumber_pic.a`, it manages to find a `.so` version from hercules 3.x...

**Comment by Rhialto** [comment]
It looks like almost exactly 2 years ago I also worked on getting hyperion to build on NetBSD. Here is a fork of the repo from then: https://github.com/Rhialto/hyperion
```
| * 3886f5bb (HEAD -> rhialto/master, origin/rhialto/master) Force libtool to accept .a files to link into shared libs.
| * 73498d3c Add configure check for pthread_rwlockattr_setpshared().
| * 7b3cca1d Add configure check for pthread_getcpuclockid().
| * 3074d936 Add "Hard-coded NetBSD-specific features and options..."
| * 661d3010 Treat NetBSD same as FreeBSD.
| * e8c20aa1 More careful use of includes when testing for further features.
| * 9efbcb8e This looks like a typo (variable does not match header file name)
|/  
* 68b03c02 CPU Timer Fix Stopping 2nd Level VM ABENDs
```
I am incorporating the following changes in my current patch set witch changes where needed. @Fish-Git  This group you may want to consider to cherry-pick or otherwise incorporate, as far as you haven't done so already.
-  73498d3c Add configure check for pthread_rwlockattr_setpshared().
- 3074d936 Add "Hard-coded NetBSD-specific features and options..."
- 661d3010 Treat NetBSD same as FreeBSD.
- e8c20aa1 More careful use of includes when testing for further features.
- 9efbcb8e This looks like a typo (variable does not match header file name)

This one I did totally different now:

- 3886f5bb Force libtool to accept .a files to link into shared libs.

This seems no longer relevant:

- 19b74bc4 Get rid of one class of the many many many many warnings from gcc 4.8.4:
- 7b3cca1d Add configure check for pthread_getcpuclockid().

**Comment by Rhialto** [comment]
The package as it is now builds ok (without disabling IPv6; the equivalent of e8c20aa takes care of that) and I tried it with a SixPack 1.3 beta, and that worked!
Hopefully I'm now finished (well, if I want to tackle tun/tap stuff I'm not, since I totally ignored that). Some final(?) remarks:
- I had to remove the `-march=xxxx` compiler option. For packaging purposes, binaries built on a build host should work on other cpu variants. When I ran hercules on my laptop, it immediately crashed with an illegal instruction. After I fixed that, it ran fine with a SixPack 1.3.
- https://github.com/NetBSD/pkgsrc-wip/tree/master/sdl-hercules-hyperion

**Comment by Fish-Git** [comment]
> > Bill Lewis wrote:
> > > util/bldlvlck doesn't work. On NetBSD, Perl is installed by default in a different place than the script is expecting.
> > 
> > 
> > I don't understand.
> 
> I meant the Perl itself that runs the bldlvlck script. It's pointed to by the shebang line:
> 
> `#!/usr/bin/perl -w`
> 
> On this NetBSD, Perl is at `/usr/local/bin/perl`
> 
> I believe one fix for that is to use:
> `#! /bin/sh`
> `PERLPROG=perl`
> `which ${PERLPROG} > /dev/null`
> `if [ "$?" -ne 0 -a -x /usr/local/bin/perl ] ; then`
> ` PERLPROG=/usr/local/bin/perl`
> `fi`
> `exec ${PERLPROG} -w -x $0 "$@"`
> `#! perl -w`
> 
> That works on NetBSD 9.0 and Debian 10 (all I've tested so far).

Committed! &nbsp;(2f51035436f29c7370b1c25b3d482af3b15bb287) &nbsp;_(which FYI I gave you full credit for)_

**Comment by Fish-Git** [comment]
Rhialto wrote (regarding the external packages):

> What's a bit weird is how the name of the build directory is relevant. I can see how it fits with the Visual Studio build but it is of course un-Unixy, and it causes some trouble later on.

Someone else designed that. I just kept it (carried it over) into my version.

> As far as I could tell, the "bitness" (32 or 64) of the build makes no difference, at least on Unix. If it were important, it is something that should be detected by cmake, not dictated to it.

Normally, yes, that would be true, but I believe the intent is to _always_ build _both_ a 32-bit _and_ a 64-bit version each time. Thus the cmake script is purposely written to do that.

> I kept having trouble with libtool, which didn't want to link with the libtelnet.a (etc) files. &nbsp; ... &nbsp; it seems that the libtool I have has rather strict ideas about when it accepts a library &nbsp; ...

It's not just your libtool, Rhialto, it's libtool in general (as in all libtools). It presumes (and in our specific case erroneously so) that `.a` static lib archives contain position _dependent_ code, and not position _independent_ code (PIC) like the way shared libraries use. In our case however, all of the external packages are _always_ built with the `-fPIC` option because we know the code is going to be eventually linked into a Hercules shared library. Thus such "helpful warnings" by libtool can be safely ignored.

If libtool is _preventing_ you from successfully building Hercules because of it however, then that's a bug _(in libtool, not Hercules!)_ that needs to be reported and fixed (although we _might_ be able to work around it, I don't know).

> Then everything linked, except for an undefined reference in various places:
> 
> ```
> /home/tmp/wip/sdl-hercules-hyperion/work.x86_64/hyperion-Release_4.2.1/.libs/libhercu.so: undefined reference to `pthread_rwlockattr_setpshared'
> ```
> 
> When I looked a bit better at the manpage for that function, I found:
> 
> ```
> BUGS
>      The pthread_rwlockattr_getpshared() and pthread_rwlockattr_setpshared()
>      functions are hidden by default since only thread shared attributes are
>      supported.
> ```
> 
> So I just commented the only call out. I hope it isn't a disaster.

Hmmm... Not sure how to fix _(or workaround)_ this particular problem. &nbsp; `:(`

Any ideas anyone?

> Finally... &nbsp; _(packaging problems related to libtool's `.la` files)_

> I need to find out how to clean that up; hopefully just an appropriate option such as `--no-static` or somesuch when creating these libraries.

I'll leave that to you then I guess.

All in all it looks like we're getting closer to cleaning all of these minor glitches up. The only thing remaining from what I can see is the adjusting (fixing/rewriting?) of our `bldlvlck` perl script to better accommodate NetBSD/FreeBSD eccentricities. Yes?

I'm going to remove "Waiting to Close" status and add "Ongoing" status instead, since there are still some minor, unresolved _(but hopefully fixable!)_ problems remaining.

**Comment by Fish-Git** [comment]
Rhialto wrote:

> I am incorporating the following changes in my current patch set with changes where needed. @Fish-Git This group you may want to consider to cherry-pick or otherwise incorporate, as far as you haven't done so already.
> 
> * [73498d3](https://github.com/SDL-Hercules-390/hyperion/commit/73498d3cc26f4535e09dbb39cacb589475b7089c) Add configure check for pthread_rwlockattr_setpshared().
> * [3074d93](https://github.com/SDL-Hercules-390/hyperion/commit/3074d936db61ca0e74ba9a0678de41a1a6c771a4) Add "Hard-coded NetBSD-specific features and options..."
> * [661d301](https://github.com/SDL-Hercules-390/hyperion/commit/661d3010bc62db0c638dd35a26e2b616e0a039d0) Treat NetBSD same as FreeBSD.
> * [e8c20aa](https://github.com/SDL-Hercules-390/hyperion/commit/e8c20aa14f4283bcbd9d8f0dddc9ee9e18d62c5b) More careful use of includes when testing for further features.
> * [9efbcb8](https://github.com/SDL-Hercules-390/hyperion/commit/9efbcb8e0c7408fc471ee8ed588feea0086e5cea) This looks like a typo (variable does not match header file name)
> 
> This one I did totally different now:
> 
> * [3886f5b](https://github.com/SDL-Hercules-390/hyperion/commit/3886f5bb01a44cb364c481f66bef809aa6ead227) Force libtool to accept .a files to link into shared libs.
> 
> This seems no longer relevant:
> 
> * [19b74bc](https://github.com/SDL-Hercules-390/hyperion/commit/19b74bc4121d23e8fd6622fd9900fc62260ac911) Get rid of one class of the many many many many warnings from gcc 4.8.4:
> * [7b3cca1](https://github.com/SDL-Hercules-390/hyperion/commit/7b3cca1d680fffa958ad8844efd9e770df551494) Add configure check for pthread_getcpuclockid().

**THANK YOU**, Rhialto!

I have reviewed your patches and have made the corresponding changes to SDL Hyperion in commit 56206efcfacc55df9f20d54c19c626fe1c236e14  _(which FYI I gave you full credit for)_.

_Hopefully_ the only issue now still remaining is the `bldlvlck` script issue. Yes?

**Comment by Fish-Git** [comment]
Rhialto wrote:

> Some final(?) remarks:
> 
> * I had to remove the `-march=xxxx` compiler option. For packaging purposes, binaries built on a build host should work on other cpu variants. When I ran hercules on my laptop, it immediately crashed with an illegal instruction. After I fixed that, it ran fine with a SixPack 1.3.

Yes, you _will_ obviously have to do that and for exactly the reason you mention: because you're building a package that to be eventually installed on a system _other than_ the one it was built on.

Our "auto-optimization" logic in `configure.ac`, as the name suggests, attempts to add reasonable default compiler optimization options under the presumption that the system it's being built on is the one it will also be run on. In your "packaging" case however, this is of course not true. Thus the need to override it.

I'm not familiar with how to build packages _(I'm barely familiar with Linux!)_ but, we do provide a `./configure` option specifically designed to allow one to easily override the defaults: the _**`--enable-optimization="my opts"`**_ option:

<pre>
[fish@centos-64 hercules-0]$ ./configure --help=short
Optional Features:
  --disable-option-checking  ignore unrecognized --enable/--with options
  --disable-FEATURE       do not include FEATURE (same as --enable-FEATURE=no)
  --enable-FEATURE[=ARG]  include FEATURE [ARG=yes]

[...]

<b><i>  --enable-optimization=yes|no|FLAGS
                          enable automatic optimization, or specify flags</i></b>

[...]

  FFLAGS      Fortran 77 compiler flags
  HQA_DIR     directory of optional "hqa.h" header

Use these variables to override the choices made by `configure' or to help
it to find libraries and programs with nonstandard names/locations.

Report bugs to the package provider.
[fish@centos-64 hercules-0]$
</pre>

As I said I don't know how you would incorporate that into your (automated?) package building procedure, but it's there if you can use it and is better than having to manually fix things.

_&lt;shrug&gt;_

Just wanted to pass that on.

**Comment by Rhialto** [comment]
Thanks for updating Hercules! At the next release I can remove some of the patches I put in pkgsrc so far, which is great.
Coming back to some previous remarks (there is not really anything here anymore to change, but just to round off things, so to speak).

The 32/64 bit build thing surely is more of a Windows thing, where it was (still is??) much more common to run 32-bit software on 64-bit capable CPUs or operating systems. There are also some Linux distributions that have separate `lib64` directories, but I think that most of those generally don't have any or many 32 bits libraries in the 32-bit directory. For myself, using NetBSD, I don't think I've ever compiled anything 32-bit to use on my DEC Alpha or later the AMD64. I'm not even sure how to do it; I suspect it is basically a cross-compile operation. Hence my desire to ignore the bitness, which I managed to do.

The `--enable-optimization=no` is a bit too drastic I think. I do want optimization, just not of the kind that is cpu-model specific. I see I could specify the exact flags I want; maybe that is good enough, as long as the configure script always sets the same flags anyway.

The `bldlvlck` script personally doesn't bother me. I ran it once by hand, looked at the result, and that was it. I'm not going to run it as part as an automated build. The pkgsrc recipe can already contain minimum required versions of dependencies and in that context it makes more sense to use that facility.

Regards and thanks! I'm closing the issue now; if something comes up later, I'll open another one.

**Comment by Fish-Git** [comment]
Hi Rhialto!

That's good news!  I'm glad I could help. 

I'm also happy to see I've resolved enough of your issues that you felt comfortable with closing the GitHub Issue. Thank you for that!

As you also say, if there's anything else that comes up later, be sure to let us know.

Thanks again and take care.

Enjoy your personal mainframe! &nbsp; `:)`
